core/process.py

```diff
@@ -969,20 +995,42 @@ def WhenDone(self, pid, status):
     self.state = job_state_e.Done
     if self.parent_pipeline:
       self.parent_pipeline.WhenDone(pid, status)
+    else:
+      self.job_state.MaybeTakeTerminal()
 
   def RunProcess(self, waiter, why):
     # type: (Waiter, trace_t) -> int
     """Run this process synchronously."""
     self.StartProcess(why)
+    # ShellExecutor might be calling this for the last part of a pipeline.
+    if self.parent_pipeline is None:
+      # QUESTION: Can the PGID of a single process just be the PID?  i.e. avoid
+      # calling getpgid()?
```

Comment:
Definitely. I think I an old version of this patch actually had a conditional here, but I figured it was useless to have an extra one unless we're worried about the overhead of an extra syscall